cache/cache.go

```diff
@@ -0,0 +1,119 @@
+package cache
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+)
+
+const NodeSize = 32
+
+type LayerFactory func(layer uint) LayerReadWriter
+
+type Cache struct {
+	layers         map[uint]LayerReadWriter
+	Hash           func(lChild, rChild []byte) []byte
+	layerFactories []LayerFactory
+}
+
+func NewCacheWithLayerFactories(layerFactories []LayerFactory) *Cache {
+	return &Cache{
+		layers:         make(map[uint]LayerReadWriter),
+		layerFactories: layerFactories,
+	}
+}
+
+type LayerReadWriter interface {
+	LayerReader
+	LayerWriter
+}
+
+type LayerReader interface {
+	Seek(index uint64) error
+	ReadNext() ([]byte, error)
+	Width() uint64
+}
+
+type LayerWriter interface {
+	io.Writer
+}
+
+func (c *Cache) GetLayerReader(layer uint) LayerReader {
+	return c.layers[layer]
+}
+
+func (c *Cache) GetLayerWriter(layer uint) LayerWriter {
+	layerReadWriter, found := c.layers[layer]
+	if !found {
+		layerReadWriter = c.fillLayerIfNeeded(layer)
+	}
+	return layerReadWriter
+}
+
+func (c *Cache) LeafReader() LayerReadWriter {
+	return c.layers[0]
```

Comment:
what if there are no layers?

Reply:
This method doesn't exist anymore, but there's:
```golang
func (c *Reader) GetLayerReader(layerHeight uint) LayerReader {
	return c.layers[layerHeight]
}
```

So the answer is that the `layers` map is initialized in the constructor and requesting a non-existent layer is a legit scenario that happens normally. If you try to read from a map with a key that doesn't exist you get the zero value, `nil` in this case.